### ISSUE TITLE
Is there a way to contact you?

### ISSUE THREAD

**Issue by dhruvbird** [kbrodt/yandex-cup-2023-neuroswipe#1]
Just wanted to chat and learn something from you about specific experiment results you might have run along the way based on the repository code contents.

**Comment by kbrodt** [comment]
You may ask here

**Comment by dhruvbird** [comment]
Thanks! I'll put down some high level questions here. 

1. How big was the training data (Ie how many gestures?)
2. Did you perform any data augmentation? It seems like you didn't based on the code but wanted to check just in case you did. 
3. Do you recall the accuracy of the raw model  using just greedy decoding? Ie no beam search and no edit distance post processing. 
4. I saw you have code for using a transformer model too. Did you try that? What were the results? 
5. How many parameters was your final model and how was it split between encoder and decoder?

**Comment by kbrodt** [comment]
1. ~6M samples. If we drop duplicated sequences by characters, we get 2.7M. But I don't know how many are "real" gestures.
2. I add gaussian noise to input curves, it prevents from overfitting after ~7 epochs
3. Without beam search the score is about 0.89. Edit distance doesn't improve score much, the difference is ~0.0001, so it's nothing
4. It is hard to train the transformer models, so in my experiments I get the score around 0.8
5. Encoder consist of one bi-directional LSTM unit with 2048 hidden size. The same for decoder. The embedding size for encoder is 4x12x128 array, and for decoder is #(characters)x128, where #(characters) ~35 if I recall correctly. In my experiments LSTM with 256 hidden size has score 0.89 with and 0.89 without beam search.

**Comment by dhruvbird** [comment]
> ~6M samples. If we drop duplicated sequences by characters, we get 2.7M. But I don't know how many are "real" gestures. 

When you say duplicated sequences do you mean the sequence of letters formed by taking each touch point and converting it into a letter by finding the character on the keyboard closest to that letter? 

Q. Did you remove the duplicate sequences or did you keep them? 

When you say you don't know how many are real sequences, what do you mean? Do you suspect that some are created artificially or that some are duplicated or that some are just bad data? 


Did you do any sort of data cleaning before training the model?

**Comment by kbrodt** [comment]
> When you say duplicated sequences do you mean the sequence of letters formed by taking each touch point and converting it into a letter by finding the character on the keyboard closest to that letter?

Yes, exactly, converting each touch points to buttons with letter.

> Did you remove the duplicate sequences or did you keep them?

At the beginning of the experiments, I removed them, but then left them as is. They play a role as augmentations, representing the same word but with a slightly different gesture curve.

> When you say you don't know how many are real sequences, what do you mean? Do you suspect that some are created artificially or that some are duplicated or that some are just bad data?

Intuitively, the number of clusters the gestures representing the words must be not greater then the number of the words. So, I mean by "real" gestures the number of such clusters. "Real" is not good word to describe what I mean :) It's better to use number of representatives.

We may say that two gestures are almost the same for the same word.

> Did you do any sort of data cleaning before training the model?

No

**Comment by kbrodt** [comment]
You may also see the [slides](https://github.com/kbrodt/yandex-cup-2023-neuroswipe/files/14736147/pres.pdf) and [video presentation](https://youtu.be/mlCQUycDTtQ?t=1483) (everything is in Russian, but YouTube generates automatic English transcription).

**Comment by dhruvbird** [comment]
1. Thanks for answering my questions - I have more - I hope you continue answering them :) 
2. Thanks for the slides and video link - please could you add them to the main README.md? I think they are very valuable. I was able to use English subtitles and understand probably 50% of what you said - which is more than I knew before - so that's awesome! 
3. I wasn't able to follow too many of the questions because they didn't translate well, but I saw that you got the question about transformer from the audience too. 

> Intuitively, the number of clusters the gestures representing the words must be not greater then the number of the words. So, I mean by "real" gestures the number of such clusters. "Real" is not good word to describe what I mean :) It's better to use number of representatives.

What do you mean by "clusters" here? Did you do any sort of clustering using some traditional ML techniques like KNN or such? Sorry this part is still not clear to me. 


Also, I'm putting down some stuff I learned from the video and slides based on what I understood for verification from you. Please could you verify and/or correct my understanding? 

#### For the input pre-processing: 

1. For the input, you "quantized" the input X/Y space into a 4x8 grid, with 4 rows and 8 columns, each having equal width and equal height (based on the keyboard dimensions or maybe normalized to 0..1). 
2. From there you considered this to be an embedding table with 32 input tokens, and each token having a feature dimension of 128. 
3. Instead of picking just the token corresponding to the square where a touch point lands, you picked 4 closest neighbouring squares (i.e. elements from the embedding table) and averaged their feature vectors and used that as the input feature vector for that specific point. 
4. Specifically, you **did not** perform any sort of weighted averaging of the feature vectors at those 4 locations based on the distance of the touch point from the center of those 32 (4x8) rectangles. 
5. If a touch point fell on the edges, then you would average just 2 or 3 feture vectors and not 4 or did you stll pick the closest 4? 

#### For validation: 

1. You claimed that removing duplicates from the validation set resulted in a lower validation accuracy - do you recall what that number was on your final solution? 
2. Intuitively, it makes sense that in any language there would be some common words and if the model was trained on biased data (for example, in English, words like `the`, `one`, `for`, `that`, etc... are very very common) the would be more likely to predict those words than other words. Do you recall what the maximum ratio between the most and least common words in the dataset was? For example, in English, one can imagine a 700:1 ratio for words like `the` v/s `henceforth`. 
3. I gather that the validation set in the private evaluation contained duplicates, and hence you continued to perform validation on the validation set with duplicates - is that an accurate statement? 

#### Out of vocabulary words: 

Since your vocabulary already had 500k words, this is probably a long shot for me to even ask, but did you try to generate some gestures for out of vocabulary words like names of people that don't occur in the training data set and see how well the model generalized on those swipe gesture inputs? 

#### Model size: 

You mentioned that the model was 120M (at least that's what the translated captions said on youtube) - I assume you mean size and not number of parameters. This would suggest that the model has 30M parameters - is that correct? I'm assuming this includes the input and output embeddings as well. 

1. Input embedding: 4x8x128 = 4k parameters 
2. Output embedding: 128x32 = 4k parameters 
3. Rest: (encoder + decoder + attention).

**Comment by kbrodt** [comment]
> What do you mean by "clusters" here? Did you do any sort of clustering using some traditional ML techniques like KNN or such? Sorry this part is still not clear to me.

The word "the" has similar gesture curves. All the curves for the word "the" belongs to the same class/cluster. Naïvely, one may compute the number of such clusters as a vocabulary size, but the dataset is not clean (there are typos like "thhe", "th", etc for the same word "the"). Another way to calculate the number of clusters is clustering algorithms on curvers, which I did not use here. It would be intersting to know how many different gestures people make.

> For the input pre-processing:

Yes, the idea is correct. Here are some clarifications.

Most participants replace the input curve to the sequence of letters, as you mentioned above:

> "the sequence of letters formed by taking each touch point and converting it into a letter by finding the character on the keyboard closest to that letter"

The drawback of this method is that the letter "W" knows nothing about the letters "Q" or "E", which are close in the English keyboard layout. To model this proximity I don't discretize the curve and stay in continuous space. I model the keyboard as a `m x n` grid (or image) with `d` channels (`4x12x128` in my experiments, `4x8` is typo). To get the embedding for coordinate `(x, y)` I sample from the grid using [bilinear interpolation](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html). It's weighted sum of closest pixels.

> For validation:

I validated locally. The score with duplicated samples is 0.88 and without -- 0.74. The correlation with the leaderboard test set is high, so I'm assuming the same result without duplicate samples in the set.
If you mean what is the score on test set of the model trained without duplicated samples, I have not sumbitted such model. I computed the score on local validation of such model and get around 0.88, almost the same as training the model with duplicated samples.

> 2

No, I didn't compute this ratio.

> 3

At first I assumed that there were no duplicate samples in the test set (in the sense above). But then I realized that they are. So yes, the statement is correct.

> Out of vocabulary words:

My approach is very simple and there are a lot of stuff to improve the model. As you said one may generate new words to improve the robustness. I didn't do that.

> Model size:

The captions are wrong :)

The model size is 2 LSTM units with 2048 hidden dimensions (9.5M encoder + 9.5M decoder) and 2 embedding arrays with sizes `4x12x128` (6144) and `~35x128` (~4480) +attention (8.4M)

**Comment by dhruvbird** [comment]
> but the dataset is not clean (there are typos like "thhe", "th", etc for the same word "the"). 

Okay, I got it! This is interesting. So, basically, the same (or very similar gesture) results in mispellings of the same underlying word, `the` (for example). 

> Another way to calculate the number of clusters is clustering algorithms on curvers, which I did not use here. It would be intersting to know how many different gestures people make.

Yeah, even I've wondered this about the data. For example, in English, the word pairs (`god`, `good`), (`to`, `too`), etc... have basically the same gesture, and it would be nice to be able to take the gesture for one of the words, and back-prop/train the model for both words. That said, there could be some minor variations in speed or time spent on the letter `o` (in the examples above) that the model can learn, and then this technique would not be as effective. 

> The drawback of this method is that the letter "W" knows nothing about the letters "Q" or "E", which are close in the English keyboard layout. To model this proximity I don't discretize the curve and stay in continuous space. I model the keyboard as a m x n grid (or image) with d channels (4x12x128 in my experiments, 4x8 is typo). To get the embedding for coordinate (x, y) I sample from the grid using [bilinear interpolation](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html). It's weighted sum of closest pixels.

Got it! thanks - this is super helpful! Q. Would you have liked to (for example) sample a 3x3 grid in 2d if there was an option to do so instead of a 2x2 grid? I guess you could have done it using the 3d version of grid_sample. I just wanted to know if it was an intentional decision to use a 2x2 sampling grid instead of something else or was it just a matter of convenience since grid_sample supports just that sampling variant for 4d input tensors. 

> If you mean what is the score on test set of the model trained without duplicated samples, I have not sumbitted such model. I computed the score on local validation of such model and get around 0.88, almost the same as training the model with duplicated samples.

Yes, that's what I was asking - thanks for the answer and the numbers! 

Thanks for providing details regarding the model size.

**Comment by dhruvbird** [comment]
I have some more questions below. 

#### Label smoothing: 

Did you perform any experiment to determine the impact of label smoothing? i.e. did you observe any meaningful impact of using or not using label smoothing? 

#### Augmentation: 

> I add gaussian noise to input curves, it prevents from overfitting after ~7 epochs

I'm trying to understand if you generated gaussian noise independently for each sample point, or if you generated a single X and Y gaussian noise component and applied it to all the points in the input gesture? I ask because intuitively, it feels like applying noise independently to each point would make the gesture pattern very **jagged** and rough, and it would no longer be representative of an actual input gesture. I was not able to locate the code for this even though I see an argument called `augs`. Maybe I just missed it. 

#### Optimizer: 

I see in your code that you tried using 3 optimizers, Adam, AdamW (wd=1e-2) and SGD (wd=1e-2). Do you recall the results of using these optimizers or different weight decays? Intuitively it feels like `1e-2` is a fairly high value of weight decay. Did you try other values too? 

#### Scheduler: 

Did the choice of scheduler make any difference to the final model performance or to the train time for the model to converge? 

#### Learning Rate: 

Q. Did you use the same LR of `5e-4` for all optimizers or did you increase the LR for SGD (if you ran any experiments with SGD that is)?

**Comment by kbrodt** [comment]
> Would you have liked to (for example) sample a 3x3 grid in 2d if there was an option to do so instead of a 2x2 grid? I guess you could have done it using the 3d version of grid_sample. I just wanted to know if it was an intentional decision to use a 2x2 sampling grid instead of something else or was it just a matter of convenience since grid_sample supports just that sampling variant for 4d input tensors.

Basicaly, each "pixel" of the grid correspond to button on a keyboard and acts as an embedding for this button. Unlike regular character embeddings, I need the proximity of the button embeddings, and `grid_sample` does exactly what I need: averages nearest embeddings by `(x, y)` coordinates. Therefore, it is better to have a grid size that matches the number of buttons on the keyboard.

> Label smoothing:

There was no difference, but I left it because this is common in nlp to use label smoothing :)

> Augmentation:

Agree, that we need to use more sophisticated noise injection, but I use [independent for each point](https://github.com/kbrodt/yandex-cup-2023-neuroswipe/blob/main/src/dataset.py#L197) and gestures become jagged (controled by variance of gaussian). Without this noise the learning curve degrades after 5-7 epochs on validation. With such stupid noise I was able to train model more epochs and get slightly better score (not signifant).

> Optimizer:

No, I used only AdamW with default parameters (`wd=1e-2`) for LSTM.
It's only when I was trying to train a transformer I tried SGD, but AdamW works much better.

> Learning Rate:

For AdamW I used [`lr=1e-3`](https://github.com/kbrodt/yandex-cup-2023-neuroswipe/blob/main/run.sh#L20) with a warm-up of 500 steps. But I didn't see any difference with or without warm-up.

When I tried to train the transformer I used Adam, AdamW, SGD with different learning rates (from "1e-5" to "1e-3") and didn't manage to train better than LSTM, even worse.

**Comment by kbrodt** [comment]
> Scheduler:

No difference in performence or speed based on validation. By default I use only warm-up. In my experiments it helps only for finetuning the pretrained models on similar tasks.

**Comment by kbrodt** [comment]
> independent for each point

actually you force the model to denoise

**Comment by dhruvbird** [comment]
> actually you force the model to denoise 

Good point, I didn't consider the denoising angle.

**Comment by dhruvbird** [comment]
Q. Did you use the timestamp during training at all? In your README.md or the videos/slides, I didn't see any mention of the use of timestamp.

**Comment by dhruvbird** [comment]
btw, thanks once again for patiently answering all my questions! :)

**Comment by kbrodt** [comment]
> Did you use the timestamp during training at all? In your README.md or the videos/slides, I didn't see any mention of the use of timestamp.

No. The speed can be used for future research.

**Comment by dhruvbird** [comment]
@kbrodt we just published something related. https://www.grammarly.com/blog/engineering/deep-learning-swipe-typing/ 
tl;dr is that the model architecture itself is very similar, but significant differences in the input features.
